go.mod

```diff
@@ -37,6 +37,7 @@ require (
 
 require (
 	github.com/DATA-DOG/go-sqlmock v1.5.0
+	github.com/ThinkChaos/parcour v0.0.0-20230418015731-5c82efbe68f5
```

Comment:
This is a module I wrote with the goal of making concurrency patterns reusable and hard to misuse.
The module is not really complete, so I just did `go get` with a specific commit.

It is fully documented, so hopefully that can help make things clear. I'll also post a comment to explain the basics in the hosts file resolver.

Reply:
I like the idea to use some "high-level" library for concurrency to reduce some boiler plate code. How does your lib differ from https://github.com/sourcegraph/conc or similar libraries? I'm not against your implementation, it is only the question about support/maturity/features etc

Reply:
I didn't find that module when I searched for existing implementations. I searched using the godoc website and GitHub, I guess I should've used a search engine too.  

A couple differences I noticed:
- conc's pools are all different types with slightly different APIs and aren't composable.  
  So no single interface can cover them all, meaning each function has to choose a flavor. Abstractions using a pool must choose for the users which type to use. See `Stream`: can't use it if you need to return errors.
  It also means that they have to implement each combination, which is not maintainable. See: `ContextPool`, `ResultContextPool`, `ResultPool`, etc. If you want to both use a context, and return an error, you can't without writing a new pool.
- conc's pools don't have a parent/child relationship, so compared to parcour, you loose the automatic error propagation to parents
- conc's pools are not well integrated with contexts
- conc's pools are not safe for concurrent use.  
  Given that it's based on `sync.WaitGroup`, just like parcour, it would be nice to be able to call `Go` and `Wait` concurrently
- parcour has `Close` and `Wait`, conc only `Wait`.
  Pro for conc: a single method is easier to understand.
  But I think it's also easier to misuse: you can't just `defer Close` and be sure you're not leaking tasks.

So I'm not convinced on conc's abstractions. The main things being they miss nesting and some of the failure propagation which are both a core part of structured concurrency. And the pools are not composable and create fractured and less reusable APIs.  
Hopefully I don't sound too harsh, I think the `iter` and `panics` packages have some interesting ideas, but globally I'm not sold.  
I'll look for other libraries to try and find something I like more.

Reply:
If you think it makes sense to make a separate implementation for this and it provides additional features we need -> why not?